### PR TITLE
Avoid overflow for `fixed_point` `cudf::cast` and performance optimization

### PR DESCRIPTION
**Pull request by codereport**
This resolves https://github.com/rapidsai/cudf/issues/9000.

When using `cudf::cast` for a wider decimal type to a narrower decimal type, you can overflow. This PR modifies the code path for this specific use case so that the "rescale" happens for the type cast. A small perf improvement was added when you have identical scales to avoid rescaling.

CI depends on https://github.com/rapidsai/cudf/pull/9766

### CODE REVIEW

File: cpp/src/unary/cast_ops.cu [hunk 1]
```diff
@@ -305,28 +305,38 @@ struct dispatch_unary_cast_to {
...
+                        device_cast<SourceDeviceT, TargetDeviceT>{});
+
+      return output;
+    };
+
+    if (input.type().scale() == type.scale()) return casted();
+
+    if constexpr (sizeof(SourceDeviceT) < sizeof(TargetDeviceT)) {
+      // device_cast BEFORE rescale when SourceDeviceT is < TargetDeviceT
+      return detail::rescale<TargetT>(*casted(), scale_type{type.scale()}, stream, mr);
```
Comment (ttnghia) [inline cpp/src/unary/cast_ops.cu:334]: Here you access the content of a temporary object (pointer), which can lead to UB.
How about this:
```
auto const casted = [&] {} (); // evaluate and store.
return ........(*casted, ......);
```

Comment (codereport) [reply cpp/src/unary/cast_ops.cu]: Done.

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/unary/cast_tests.cpp [hunk 2]
```diff
@@ -1003,3 +1003,16 @@ TYPED_TEST(FixedPointTests, Decimal128ToDecimalXXWithLargerScaleAndNullMask)
 
   CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, result->view());
 }
+
+TEST_F(FixedPointTestSingleType, Int32ToInt64Convert)
```
Comment (ttnghia) [inline cpp/tests/unary/cast_tests.cpp:1007]: Nit: I'm not sure if we also need to have another test for casting int64 to int32.

Comment (codereport) [reply cpp/tests/unary/cast_tests.cpp:1007]: This is the test that failed from the issue, so it was not covered previously.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Comment by codereport** [comment]
@gpucibot merge

**Pull request merged**